File: frontend/src/api/attendance.ts
import api from './client';

export interface AttendanceMethodResponse {
  method: 'qr_code' | 'face_id';
  allow_company_override: boolean;
  is_company_override: boolean;
}

export interface QrTokenResponse {
  token: string;
  expires_at: string;
  scan_url: string;
}

export interface AttendanceRecord {
  id: string;
  company_id: string;
  employee_id: string;
  check_in_at: string;
  check_out_at: string | null;
  method: 'qr_code' | 'face_id' | 'manual';
  status: 'present' | 'late' | 'absent' | 'half_day';
  latitude: number | null;
  longitude: number | null;
  checkout_latitude: number | null;
  checkout_longitude: number | null;
  notes: string | null;
  qr_token_id: string | null;
  hours_worked: number | null;
  overtime_hours: number | null;
  is_outside_geofence: boolean | null;
  created_at: string;
  updated_at: string;
}

export interface AttendanceRecordWithEmployee extends AttendanceRecord {
  employee_number: string;
  full_name: string;
  department: string | null;
}

export interface AttendanceListQuery {
  employee_id?: string;
  date_from?: string;
  date_to?: string;
  status?: string;
  method?: string;
  page?: number;
  per_page?: number;
}

export interface PaginatedResponse<T> {
  data: T[];
  total: number;
  page: number;
  per_page: number;
  total_pages: number;
}

// ─── Method Control ───

export function getAttendanceMethod(): Promise<AttendanceMethodResponse> {
  return api.get('/attendance/method').then(r => r.data);
}

export function getPlatformAttendanceMethod(): Promise<{ method: string; allow_company_override: boolean }> {
  return api.get('/admin/platform/attendance-method').then(r => r.data);
}

export function setPlatformAttendanceMethod(method: string, allow_company_override: boolean) {
  return api.put('/admin/platform/attendance-method', { method, allow_company_override }).then(r => r.data);
}

export function setCompanyAttendanceMethod(method: string | null) {
  return api.put('/attendance/company-method', { method }).then(r => r.data);
}

// ─── QR Token ───

export function generateQrToken(): Promise<QrTokenResponse> {
  return api.post('/attendance/qr/generate').then(r => r.data);
}

// ─── Check In / Out ───

export function checkInQr(token: string, latitude?: number, longitude?: number): Promise<AttendanceRecord> {
  return api.post('/attendance/check-in/qr', { token, latitude, longitude }).then(r => r.data);
}

export function checkInFaceId(credential_id: string, assertion: unknown, latitude?: number, longitude?: number): Promise<AttendanceRecord> {
  return api.post('/attendance/check-in/face-id', { credential_id, assertion, latitude, longitude }).then(r => r.data);
}

export function checkOut(latitude?: number, longitude?: number): Promise<AttendanceRecord> {
  return api.post('/attendance/check-out', { latitude, longitude }).then(r => r.data);
}

// ─── Portal ───

export function getMyTodayAttendance(): Promise<{ record: AttendanceRecord | null }> {
  return api.get('/attendance/my/today').then(r => r.data);
}

export function getMyAttendance(params?: AttendanceListQuery): Promise<PaginatedResponse<AttendanceRecord>> {
  return api.get('/attendance/my', { params }).then(r => r.data);
}

// ─── Admin ───

export function getAttendanceRecords(params?: AttendanceListQuery): Promise<PaginatedResponse<AttendanceRecordWithEmployee>> {
  return api.get('/attendance/records', { params }).then(r => r.data);
}

export function createManualAttendance(data: {
  employee_id: string;
  check_in_at: string;
  check_out_at?: string;
  status?: string;
  notes?: string;
}): Promise<AttendanceRecord> {
  return api.post('/attendance/manual', data).then(r => r.data);
}

export function updateAttendanceRecord(id: string, data: {
  check_in_at?: string;
  check_out_at?: string;
  status?: string;
  notes?: string;
}): Promise<AttendanceRecord> {
  return api.put(`/attendance/records/${id}`, data).then(r => r.data);
}

File: backend/src/models/attendance.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

// ─── Platform Setting ───

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct PlatformSetting {
    pub key: String,
    pub value: String,
    pub updated_at: DateTime<Utc>,
    pub updated_by: Option<Uuid>,
}

// ─── QR Token ───

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct AttendanceQrToken {
    pub id: Uuid,
    pub company_id: Uuid,
    pub token: String,
    pub expires_at: DateTime<Utc>,
    pub used: bool,
    pub created_at: DateTime<Utc>,
}

// ─── Attendance Record ───

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct AttendanceRecord {
    pub id: Uuid,
    pub company_id: Uuid,
    pub employee_id: Uuid,
    pub check_in_at: DateTime<Utc>,
    pub check_out_at: Option<DateTime<Utc>>,
    pub method: String,
    pub status: String,
    pub latitude: Option<f64>,
    pub longitude: Option<f64>,
    pub checkout_latitude: Option<f64>,
    pub checkout_longitude: Option<f64>,
    pub notes: Option<String>,
    pub qr_token_id: Option<Uuid>,
    pub created_by: Option<Uuid>,
    pub hours_worked: Option<rust_decimal::Decimal>,
    pub overtime_hours: Option<rust_decimal::Decimal>,
    pub is_outside_geofence: Option<bool>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// Extended record that includes employee details for admin list views
#[derive(Debug, Serialize, sqlx::FromRow)]
pub struct AttendanceRecordWithEmployee {
    pub id: Uuid,
    pub company_id: Uuid,
    pub employee_id: Uuid,
    pub employee_number: String,
    pub full_name: String,
    pub department: Option<String>,
    pub check_in_at: DateTime<Utc>,
    pub check_out_at: Option<DateTime<Utc>>,
    pub method: String,
    pub status: String,
    pub latitude: Option<f64>,
    pub longitude: Option<f64>,
    pub checkout_latitude: Option<f64>,
    pub checkout_longitude: Option<f64>,
    pub notes: Option<String>,
    pub hours_worked: Option<rust_decimal::Decimal>,
    pub overtime_hours: Option<rust_decimal::Decimal>,
    pub is_outside_geofence: Option<bool>,
    pub created_at: DateTime<Utc>,
}

// ─── Requests ───

#[derive(Debug, Deserialize)]
pub struct CheckInQrRequest {
    /// The raw token value from scanning the QR code
    pub token: String,
    pub latitude: Option<f64>,
    pub longitude: Option<f64>,
}

#[derive(Debug, Deserialize)]
pub struct CheckInFaceIdRequest {
    /// Raw ID of the WebAuthn credential used (from webauthn assertion)
    pub credential_id: String,
    /// Client assertion data (base64url encoded)
    pub assertion: serde_json::Value,
    pub latitude: Option<f64>,
    pub longitude: Option<f64>,
}

#[derive(Debug, Deserialize)]
pub struct CheckOutRequest {
    pub latitude: Option<f64>,
    pub longitude: Option<f64>,
}

#[derive(Debug, Deserialize)]
pub struct ManualAttendanceRequest {
    pub employee_id: Uuid,
    pub check_in_at: DateTime<Utc>,
    pub check_out_at: Option<DateTime<Utc>>,
    pub status: Option<String>,
    pub notes: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct AttendanceListQuery {
    pub employee_id: Option<Uuid>,
    pub date_from: Option<String>,
    pub date_to: Option<String>,
    pub status: Option<String>,
    pub method: Option<String>,
    /// Page number (1-based, default 1)
    pub page: Option<i64>,
    /// Items per page (default 50, max 200)
    pub per_page: Option<i64>,
}

#[derive(Debug, Serialize)]
pub struct PaginatedAttendance<T: Serialize> {
    pub data: Vec<T>,
    pub total: i64,
    pub page: i64,
    pub per_page: i64,
    pub total_pages: i64,
}

#[derive(Debug, Deserialize)]
pub struct UpdateAttendanceRecordRequest {
    pub check_in_at: Option<DateTime<Utc>>,
    pub check_out_at: Option<DateTime<Utc>>,
    pub status: Option<String>,
    pub notes: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct SetAttendanceMethodRequest {
    /// "qr_code" or "face_id"
    pub method: String,
    /// whether admins can override the global setting
    pub allow_company_override: Option<bool>,
}

#[derive(Debug, Deserialize)]
pub struct SetCompanyAttendanceMethodRequest {
    /// "qr_code" or "face_id" or null to use platform default
    pub method: Option<String>,
}

/// Response returned when a QR token is generated
#[derive(Debug, Serialize)]
pub struct QrTokenResponse {
    pub token: String,
    pub expires_at: DateTime<Utc>,
    /// The full URL the QR code should encode (employee scans this)
    pub scan_url: String,
}

/// Response for the effective attendance method of a company
#[derive(Debug, Serialize)]
pub struct AttendanceMethodResponse {
    pub method: String,
    pub allow_company_override: bool,
    pub is_company_override: bool,
}
